evaluation/gpqa/run_infer.py

```diff
@@ -0,0 +1,455 @@
+        for handler in logger.handlers[:]:
+            logger.removeHandler(handler)
+        file_handler = logging.FileHandler(log_file)
+        file_handler.setFormatter(
+            logging.Formatter('%(asctime)s - %(levelname)s - %(message)s')
+        )
+        logger.addHandler(file_handler)
+    else:
+        logger.info(f'Starting evaluation for instance {instance.instance_id}.')
+
+    if not skip_workspace_mount:
+        logger.info(f'Process-specific workspace mounted at {workspace_mount_path}')
+
+    # ======= Run the agent on the instance =======
+    # Prepare instruction for the agent using suggested format in gpqa codebase
+    instruction = f"""
+    What is the correct answer to this question:\n
+    {instance['question']}\n
+
+    Choices:\n
+    (A) {instance['choices'][0]}\n
+    (B) {instance['choices'][1]}\n
+    (C) {instance['choices'][2]}\n
+    (D) {instance['choices'][3]}\n
+    \n\n
+
+    MOST IMPORTANT: Format your response as follows:
+    <<FINAL_ANSWER||
+    <insert correct answer here, must be one of A, B, C, D> (Please dont use any additional characters. Just the letter of the correct answer (A/B/C/D).)
+    ||FINAL_ANSWER>>
+
+    Additional Instructions:
+    - You should ONLY interact with the environment provided to you AND NEVER ASK FOR HUMAN HELP.
+    """
+
+    # NOTE: You can actually set slightly different instruction for different agents
+    instruction += AGENT_CLS_TO_INST_SUFFIX.get(agent_class, '')
+
+    # Here's how you can run the agent (similar to the `main` function) and get the final task state
+    state: State = asyncio.run(
+        main(
+            instruction,
+            fake_user_response_fn=AGENT_CLS_TO_FAKE_USER_RESPONSE_FN.get(agent_class),
+        )
+    )
+
+    # ======= Attempt to evaluate the agent's edits =======
+    # get the final message from the state history (default to None if not found)
+    final_message = next(
+        (
+            act.content
+            for act in reversed(state.history)
+            if isinstance(act, MessageAction)
+        ),
+        None,
+    )
+
+    logger.info(f'Final message generated by the agent: {final_message}')
+
+    test_result = get_test_result(final_message, instance.correct_solution)
+
+    # If you are working on some simpler benchmark that only evaluates the final model output (e.g., in a MessageAction)
+    # You can simply get the LAST `MessageAction` from the returned `state.history` and parse it for evaluation.
+    if state is None:
+        raise ValueError('State should not be None.')
+
+    # Save the output
+    output = {
+        'task_id': instance.task_id,
+        'instance_id': instance.instance_id,
+        'instruction': instruction,
+        'metadata': metadata,
+        'history': [
+            (event_to_dict(action), event_to_dict(obs)) for action, obs in state.history
+        ],
+        'error': state.error if state and state.error else None,
+        'test_result': test_result,
```

Comment:
Please also add `metrics`, see https://github.com/OpenDevin/OpenDevin/pull/2199 for example

Reply:
Yeah, pls add `metrics`, then we can get the average cost of each task which need in our final paper.